mesonbuild/interpreter.py

```diff
@@ -3616,8 +3662,32 @@ def dependency_fallback(self, name, display_name, kwargs):
             'required': required,
         }
         self.do_subproject(dirname, 'meson', sp_kwargs)
+        self.override_dependencies_from_wrap(dirname, kwargs)
         return self.get_subproject_dep(name, display_name, dirname, varname, kwargs)
 
+    def override_dependencies_from_wrap(self, dirname, kwargs):
+        # Make sure to override every dependencies this wrap provides in the
+        # case the subproject did not call meson.override_dependency().
+        # This ensures that if we fallbacked for one dependency, any other
+        # dependency provided by this subproject won't be picked from system.
+        # Also verify consistency in the case the subproject overridden the
+        # dependency with a different object than the variable from wrap file.
+        for_machine = self.machine_from_native_kwarg(kwargs)
+        provide = self.environment.wrap_resolver.get_provide(dirname)
+        for depname, varname in provide.items():
+            if not varname:
+                continue
+            d = self.subprojects[dirname].get_variable_method([varname], {})
+            identifier = dependencies.get_dep_identifier(depname, kwargs)
+            cached_dep = self.build.dependency_overrides[for_machine].get(identifier)
+            if cached_dep:
+                if d.held_object != cached_dep:
+                    m = 'Inconsistency: The dependency {!r} was overridden by another object than variable {!r} from subproject {!r}'
```

Comment:
Hmm, maybe this should be downgraded to a warning? I wanted to be stricter because we can always relax later if it cause issues, we can't go the other way around.

Reply:
I'm torn, because variables names are of course already part of the api contract of a subproject. But we move in a direction where there are less and less. For subprojects that only get used via fallback we might be quite close to get them back to implementation detail again. But with this check that stale .wrap somewhere that certainly exists will make it hard for projects to refactor. On the other hand, this only matters with .wraps that don't pin to a concrete version. So maybe it's a bit of a non-issue because of that.

Reply:
If you don't pin the version of your subproject, and that subproject break their "api" (change variable name) you'll have problems anyway as soon as you'll try to get that dependency. Better detect it early and raise exception IMHO.

Reply:
If the project does override_dependency properly you should not have any problem for the usage as fallback, should you?
Usage as classic subproject would still break.